### DISCUSSION TITLE
Support TTL of member

### DISCUSSION THREAD

**Discussion by caipengbo** [apache/kvrocks#399] [General]
## Background
Kvrocks currently supports TTL(Time to Live) at the key level. In some scenarios, Users need members to support TTL. 

#283 also raises this issue. As @git-hulk said in this issue, the implementation is not particularly difficult, but the changes may be quite large because we should add new type and there are so many additional commands to support.

Here are my design and implementation ideas that I would like to discuss with developers. I take `Hash` as an example:

##  Design
1. Add a new type, called `ExHash`
2. Based on the coding design of `Hash` Type, we encode expire time in the value of subkey:

```
        +----------+------------+-----------+-----------+
key =>  |  flags   |  expire    |  version  |  size     |
        | (1byte)  | (4byte)    |  (8byte)  | (8byte)   |
        +----------+------------+-----------+-----------+
```
```
                     +---------------+
key|version|field => |     value     |
                     |     (Nbyte)   |
                     +---------------+
```
```
                     +----------+---------------+
key|version|field => |  expire  |   value       |
                     |  (4byte) |  (Nbyte)      |
                     +----------+---------------+
```
3. In addition to commands that have the same semantics as `Hash`, `ExHash` has the following **special** commands:

| Command         | Syntax                                   | Description  |
| ----------------- | --------------------------------------------- | ------------ |
| EXHSETEX        | EXHSETEX key field value seconds | Adds a field with a second timeout to a specified ExHash. If the key does not exist, a key for the ExHash is created. If the field already existed, this command overwrites the value of the field.  |
| EXHEXPIRE        | EXHEXPIRE key field seconds   | Specifies the timeout of a field in a specified ExHash. If the field timeout already existed, this command overwrites the timeout of the field. |
| EXHEXPIREAT  | EXHEXPIREAT  key field timestamp   | Like EXHEXPIRE, just using unix timestamp. |
| EXHTTL       | EXHTTL key field                     | Get the remaining expiration time of a field in a specified ExHash. Unit: Seconds. |
| EXHPERSIST   | EXHPERSIST key field                 | Remove the timeout of a field in a specified ExHash. |

## Implement
1. Add a new `RedisType`, called `kRedisExHash`
2. Modify the `InternalKey` class and add member variables about TTL to it. When users obtain corresponding values, they need to check expire
3. In `SubKeyFilter::IsKeyExpired`, judge whether the Value expire

**Comment by caipengbo** [comment]
How do you think @KvrocksLabs/committers ?

**Reply by git-hulk** [reply]
Many thanks for @caipengbo starting the discussion and draft design here.

As mentioned previously in our slack channel, the only issue was that accessing the metadata `size` field would have the data race in the compaction filter. I and @ShooterIT agree that can sacrifice the `hlen` command. I'm not worried about the huge change if our design was simple enough, also we have many eyes to make bugs shallow.

cc @ShooterIT @Alfejik @smartlee @popunit
